core/cas-server-core-services/src/test/java/org/apereo/cas/services/AbstractServicesManagerTests.java

```diff
@@ -180,16 +177,91 @@ public void verifyServiceCanBeFoundDuringLoadWithoutCacheInvalidation() throws E
         loadingThread.join();
     }
 
+    /**
+     * Verify that a service which is <strong>not</strong> associated with the current environment ({@code current-env}
+     * in this test) is not incorrectly selected due to "promotion" through the services cache. Scenario is as follows:
+     * <ul>
+     *      <li/> Active environment is {@code current-env}.
+     *      <li/> Service {@code https://prod.test.edu} associated with {@code current-env}, low
+     *      {@code evaluationOrder}.
```

Comment:
Please use a link tag if you want to refer to a field name.

Reply:
Ditto above, this is a private field which I cannot `{@link}` without breaking Javadoc compilation, though I can update it to reference the getter instead.

Reply:
You can either use the getter, or remove the field, or use it in the descriptive text. Private implementation details cannot be javadoc'ed.

Reply:
Replaced with `{@link}` to the getter. I believe the current branch addresses all requested changes.

Reply:
Very nice, thanks. Please also sign the CLA, and then [submit PRs for all other branches](https://apereo.github.io/cas/developer/Contributor-Guidelines.html#destination-branch). I might be able to cherry pick on your behalf, if that is something you'd like me to do. Your call.

Reply:
I feel like there's less change of catastrophe if you decide where it needs to go and cherry pick. Certainly I'd be willing to do it if that puts too much work on your plate, but I'd need the list of branches to which it needs to go. I know 6.4.x and master, but I'm not sure how far back to go--don't know what has reached EOL and what has not (I'm sure this is spelled out somewhere and I'm just missing it.)

Reply:
No problem at all. I appreciate your willingness, and given the changeset is rather modest, I am sure I can do it without too much trouble.

For future reference, this should come in handy:
https://apereo.github.io/cas/developer/Maintenance-Policy.html

Reply:
Thank you. Per the maintenance policy, I would say that there is a very minor vulnerability aspect here. It would require a CAS deployment that is configured in a very specific way, and it would require specific knowledge of this issue. One *could* make the argument that it warrants back patches to any branches which are still receiving vulnerability updates.

Basically, anyone who knows about this *could* create a DoS attack of duration up to the invalidation interval of the services cache. In terms of the test, one could deny access to https://prod.test.edu by repeatedly issuing requests to https://non-prod.test.edu. Occasionally they'd get "lucky" and cause https://prod.test.edu to produce "not authorized to use CAS" errors until the service cache is purged again.

I think there's very little risk of running into this in the real world since it requires a mix of services associated with active and inactive environments and one service which is sort of a "wildcard" match that could be used to block another. But if there's a security/vulnerability angle here, that is it.

This can also be worked around by just putting service JSON files in different directories for different environments. Ultimately that's what I did to address the issue in our environment. That's a fully secure workaround from what I can see.

Reply:
I appreciate your analysis here, but this is not the right place to discuss that sort of thing, however remote it may be. If it's something you'd like to discuss with others, you should review: https://apereo.github.io/cas/developer/Sec-Vuln-Response.html

Reply:
Oh, sure, sorry about that. I'm new here. :)

Reply:
I understand. No worries :)